src/rez/utils/pip.py

```diff
@@ -0,0 +1,705 @@
+
+    if "platform" in sys_requires:
+        sys_variant_requires.append("platform-%s" % _system.platform)
+
+    if "arch" in sys_requires:
+        sys_variant_requires.append("arch-%s" % _system.arch)
+
+    if "os" in sys_requires:
+        sys_variant_requires.append("os-%s" % _system.os)
+
+    if "python" in sys_requires:
+        # Add python variant requirement. Note that this is always MAJOR.MINOR,
+        # because to do otherwise would mean analysing any present env markers.
+        # This could become quite complicated, and could also result in strange
+        # python version ranges in the variants.
+        #
+        sys_variant_requires.append("python-%s" % str(python_version.trim(2)))
+
+    return {
+        "requires": result_requires,
+        "variant_requires": sys_variant_requires + result_variant_requires,
+        "metadata": {
+            "is_pure_python": is_pure_python
+        }
+    }
+
+
+def convert_distlib_to_setuptools(installed_dist):
+    """Get the setuptools equivalent of a distlib installed dist.
+
+    Args:
+        installed_dist (`distlib.database.InstalledDistribution`: Distribution
+            to convert.
+
+    Returns:
+        `pkg_resources.DistInfoDistribution`: Equivalent setuptools dist object.
+    """
+    path = os.path.dirname(installed_dist.path)
+    setuptools_dists = pkg_resources.find_distributions(path)
+
+    for setuptools_dist in setuptools_dists:
+        if setuptools_dist.key == installed_dist.key:
+            return setuptools_dist
+
+    return None
+
+
+def get_marker_sys_requirements(marker):
+    """Get the system requirements that an environment marker introduces.
+
+    Consider:
+
+        'foo (>1.2) ; python_version == "3" and platform_machine == "x86_64"'
+
+    This example would cause a requirement on python, platform, and arch
+    (platform as a consequence of requirement on arch).
+
+    See:
+    * vendor/packaging/markers.py:line=76
+    * https://www.python.org/dev/peps/pep-0508/#id23
+
+    Args:
+        marker (str): Environment marker string, eg 'python_version == "3"'.
+
+    Returns:
+        List of str: System requirements (unversioned).
+    """
+    _py = "python"
+    _plat = "platform"
+    _arch = "arch"
+    _os = "os"
+
+    sys_requires_lookup = {
+        # TODO There is no way to associate a python version with its implementation
+        # currently (ie CPython etc). When we have "package features", we may be
+        # able to manage this; ignore for now
+        "implementation_name": [_py],
+        "implementation_version": [_py],
+        "platform_python_implementation": [_py],
+        "platform.python_implementation": [_py],
+        "python_implementation": [_py],
+
+        "sys.platform": [_plat],
+        "sys_platform": [_plat],
+
+        "os.name": [_plat, _arch, _os],
+        "os_name": [_plat, _arch, _os],
+
+        "platform.machine": [_plat, _arch],
+        "platform_machine": [_plat, _arch],
```

Comment:
Technically the platform isn't necessary here. What was the thinking behind adding the platform? I don't mind to add it, but I would prefer it to be documented I think.

Reply:
From Allan:
>The thinking was that I don't think I've ever seen an only arch-dependent
package before. If that ever makes sense then I'm happy to change. However
I get the sense that python packaging requirements are somewhat loose,
perhaps it's good to be pessimistic - ie to assume that platform.machine,
os.name etc actually implies platform dependency? What do you think?

I see what you mean. I have the feeling we shouldn't take guesses here and just do a one to one with the PEP. So basically each should only have one value. Like `os_name` should only map to `_os` (since the `os` package already requires the arch and platform package). `platform_machine` is really platform agnostic. It could weel be `x64` or `x86_64` or `x32` for example. That would be for a C extension pip package I guess.

Does that make sense?

Reply:
yeah, will go with that, if we hit issues we can always revert to the more pessimistic implementation later.